### PR TITLE
display confirmation pages after successful donation

### PR DESCRIPTION
**Pull request by KaiNissen**

### CODE REVIEW

File: src/Presentation/Presenters/AddDonationHtmlPresenter.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+class AddDonationHtmlPresenter {
+
+	private $template;
+
+	public function __construct( TwigTemplate $template ) {
+		$this->template = $template;
+	}
+
+	public function present( Donation $donation ): string {
+		return $this->template->render( [ 'donation' => $donation ] );
```
Comment (JeroenDeDauw) [inline src/Presentation/Presenters/AddDonationHtmlPresenter.php]: Like this the presenter does not add real value. Normally it should construct its own View Model. `AddDonationResponse` is already somewhat dangerous by passing the domain object to the controller code. Passing it on to the view itself is really not nice.

File: src/UseCases/AddDonation/AddDonationResponse.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+class AddDonationResponse extends ValidationResponse {
+
+	private $donation;
+
+	public function __construct( Donation $donation ) {
+		$this->donation = $donation;
+		parent::__construct();
+	}
+
+	public function getDonation(): Donation {
```
Comment (JeroenDeDauw) [inline src/UseCases/AddDonation/AddDonationResponse.php]: I'm fine (though not very happy) with returning the domain object here, though then we should be real careful with not using domain logic in the controller or putting controller or presentation related concerns into the domain object.

File: src/UseCases/AddDonation/AddDonationResponse.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+namespace WMDE\Fundraising\Frontend\UseCases\AddDonation;
+
+use WMDE\Fundraising\Frontend\Domain\Model\Donation;
+use WMDE\Fundraising\Frontend\ResponseModel\ValidationResponse;
+
+/**
+ * @license GNU GPL v2+
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ */
+class AddDonationResponse extends ValidationResponse {
```
Comment (gbirke) [inline src/UseCases/AddDonation/AddDonationResponse.php]: I'm not sure what we gain by extending `ValidationResponse`, as the constructor is made in a way that the response is always successful. Do you really need the functionality provided by `ValidationResponse`?

Comment (JeroenDeDauw) [inline src/UseCases/AddDonation/AddDonationResponse.php]: hah, I had not even noticed this inheritance :)

File: src/Presentation/Presenters/AddDonationHtmlPresenter.php (hunk 1 above)
Comment (KaiNissen) [inline src/Presentation/Presenters/AddDonationHtmlPresenter.php]: Right, changed it to accept an array of template arguments.

File: src/UseCases/AddDonation/AddDonationResponse.php (hunk 3 above)
Comment (KaiNissen) [inline src/UseCases/AddDonation/AddDonationResponse.php]: Needed some of it. I removed the inheritance.

File: src/UseCases/AddDonation/AddDonationUseCase.php [hunk 4]
```diff
@@ -177,4 +179,54 @@ private function getConfirmationMailTemplateArguments( Donation $donation, bool
...
+				'optsIntoNewsletter' => $donation->getOptsIntoNewsletter(),
+				'bankTransferCode' => $donation->getBankTransferCode(),
+				// TODO: use locale to determine the date format
+				'creationDate' => ( new \DateTime() )->format( 'd.M.Y' ),
+				// TODO: set cookie duration for "hide banner cookie"
+				'cookieDuration' => ''
+			],
+			$this->getPersonArguments( $donation ),
+			$this->getBankDataArguments( $donation )
+		] );
```
Comment (JeroenDeDauw) [inline src/UseCases/AddDonation/AddDonationUseCase.php]: This code belongs in the presenter. The fact that there is a template is an implementation detail of the presenter.

Comment (KaiNissen) [inline src/UseCases/AddDonation/AddDonationUseCase.php]: Right, moved it there.

**Comment by JeroenDeDauw** [comment]
+2 from my side when #236 is merged, except for the template, which someone else should review

**Pull request merged**